### ISSUE TITLE
curl http://127.0.0.1:5001/webui  reports "moved permanently"

### ISSUE THREAD

**Issue by ThomasFreedman** [ipfs-shipyard/ipfs-webui#1013]
**Describe the bug**
I cannot view the webui. I am testing ipfs install scripts. The first used a binary installation and I had no issues forwarding port 5001 thru ssh to view webui (ssh -L 5001:localhost:5001 ipfs_1@192.168.5.189). The second script used a go (version 1.10) based ipfs installation using ipfs-update to obtain the latest ipfs code. 

I get a 403 error trying to connect via ssh tunnel, so I logged into the ipfs server and ran curl to view webui locally. The 403 error on the remote ssh tunnel end indicates I reached the http server on the ipfs node, which is a headless server. However even trying to access the webui locally I get this result (is this an artifact of redirection or the bug?):

ipfs-node_1:~ $ curl http://127.0.0.1:5001/webui
&lt;a href="/webui/"&gt;Moved Permanently&lt;/a&gt;.

**To Reproduce**
1) install ipfs using ipfs-update (golang utility)
2) verify daemon is listening on port 5001 of server host
3) curl http://127.0.0.1:5001/webui

**Expected behavior**
I expect to see the html of the webui

**Desktop (please complete the following information):**
 - Raspbian for RPi3 (April 8th release, lite version)
 - Browser = curl version 7.52.1
 - ipfs version 0.4.19

**Comment by fphammerle** [comment]
`/webui` forwards to `/webui/`:
```sh
$ curl -I http://127.0.0.1:5001/webui
HTTP/1.1 301 Moved Permanently
Content-Type: text/html; charset=utf-8
Location: /webui/
Date: Mon, 22 Apr 2019 09:39:28 GMT
```

`/webui/` finally forwards to the webui:
```sh
$ curl -I http://127.0.0.1:5001/webui/
HTTP/1.1 302 Found
Content-Type: text/html; charset=utf-8
Location: /ipfs/QmfQkD8pBSBCBxWEwFSu4XaDVSWK6bjnNuaWZjMyQbyDub
Date: Mon, 22 Apr 2019 09:40:25 GMT
```

Have you tried `curl -I http://127.0.0.1:5001/ipfs/QmfQkD8pBSBCBxWEwFSu4XaDVSWK6bjnNuaWZjMyQbyDub/` (or similar) ?

**Comment by ThomasFreedman** [comment]
Yes. According to netstat -tulpn ipfs daemon is listening on port 5001. I save results of ips init, so I presume your suggestion is to use that key in your curl like so:

    ipfs@ipfs-node:~ $ cat .ipfs/init.log 
    initializing IPFS node at /home/ipfs/.ipfs
    generating 2048-bit RSA keypair...done
    peer identity: QmTjFtazTZc6Z5nJvT5NGuzXsoqKMhwCNB4pUw3Uja938R
    to get started, enter:

    	ipfs cat /ipfs/QmS4ustL54uo8FzR9455qaxZwuMiUhyvMcX9Ba8nUH4uVv/readme

    ipfs@ipfs-node:~ $ curl -I http://127.0.0.1:5001/ipfs/QmS4ustL54uo8FzR9455qaxZwuMiUhyvMcX9Ba8nUH4uVv/readme
    HTTP/1.1 404 Not Found
    Content-Type: text/plain; charset=utf-8
    X-Content-Type-Options: nosniff
    Date: Tue, 23 Apr 2019 01:42:22 GMT
    Content-Length: 19

Using ipfs 4.19 installed with ipfs-update. Updating right now to version 4.20. I can cat the readme file as suggested in the output of ipfs init, but no webui, even using curl on ipfs-node (localhost).

**Comment by fphammerle** [comment]
Could you run and post the output of `curl -IL http://127.0.0.1:5001/webui`
?

(To check where `/webui` finally redirects to)

**Comment by ThomasFreedman** [comment]
HTTP/1.1 301 Moved Permanently 
    Content-Type: text/html; charset=utf-8
    Location: /webui/
    Date: Tue, 23 Apr 2019 17:43:06 GMT

    HTTP/1.1 302 Found
    Content-Type: text/html; charset=utf-8
    Location: /ipfs/QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ
    Date: Tue, 23 Apr 2019 17:43:06 GMT

    HTTP/1.1 301 Moved Permanently
    Content-Type: text/html; charset=utf-8
    Location: /ipfs/QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ/
    Date: Tue, 23 Apr 2019 17:43:06 GMT

That's it, and it doesn't return to cmd prompt, but presume that's due to curl options.

**Comment by fphammerle** [comment]
So the issue isn't `HTTP/1.1 301 Moved Permanently` at `/webui`,
but no response for `/ipfs/QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ/`

`QmXc9...` is a valid hash, referring to a version of the webui.
http://ipfs.io/ipfs/QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ/

Apparently your ipfs instance won't load the data the hash refers to.

Could a firewall be blocking ipfs traffic?

**Comment by ThomasFreedman** [comment]
I thought about that. I did have a firewall installed, but disabled it, so that couldn't be the issue. The node is on a limited DSL link with only 50bps upload bandwidth, but I wouldn't have thought that should be the cause.

When you say "it won't load the data the hash refers to", isn't that hash referring to something local, like the readme? That works, as the output from ipfs init (ipfs cat /ipfs/QmS4ustL54uo8FzR9455qaxZwuMiUhyvMcX9Ba8nUH4uVv/readme
) indicates.

**Comment by fphammerle** [comment]
> When you say "it won't load the data the hash refers to", isn't that hash referring to something local, like the readme?

I am not sure whether the webui comes pre-installed or has to be fetched via ipfs.

Does `ipfs ls [webui multihash]` succeed?

```sh
$ ipfs ls QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ
Qmc5m94Gu7z62RC8waSKkZUrCCBJPyHbkpmGzEePxy2oXJ 1     _previous-versions
QmQe8KWVQhhvHsiExWTikJPjKhSCRRL5GTrpNxGskjME92 2934  asset-manifest.json
QmUVmQX9RQpn2iDJ5CKraqGN412UWAmwyvSgqB7bzYovSV 34494 favicon.ico
QmekoF89LTA1TRxvJVXKpf8L7zRYctdzQncyP85rdFeyGs 675   index.html
QmVjGNJf1MBeZxuBcy44WobcpjecUASAjyCqKzdk8Z4opa -     locales/
QmPKPhRf2FaxEPeYPa1PebmoLLENgHyq7Sd8UJSSVpnti4 317   manifest.json
QmeTJ6DsPPSFVKxW7PoKseZTQJisbYZ9iRRpVd5WW1fsND 5208  service-worker.js
QmcMqeTMZamtpWVbaY14AYwMHXLVMvLrqFTY62ccPMycXN -     static/
```

**Comment by ThomasFreedman** [comment]
No: 

    ipfs@ipfs-node:~ $ ipfs ls
    ipfs: Reading from /dev/stdin; send Ctrl-d to stop.
    ^C
    Error: Post http://127.0.0.1:5001/api/v0/ls?encoding=json&resolve-type=true&size=true&stream-channels=true: context canceled
    ipfs@ipfs-node:~ $ ipfs ls 
    ipfs: Reading from /dev/stdin; send Ctrl-d to stop.
    Error: argument "ipfs-path" is required
    ipfs@ipfs-node:~ $ ipfs ls [ipfs multihash]
    Error: invalid 'ipfs ref' path
    ipfs@ipfs-node:~ $ ipfs ls .
    Error: invalid 'ipfs ref' path

**Comment by fphammerle** [comment]
Could you run `ipfs ls QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ` ?

**Comment by ThomasFreedman** [comment]
Looks like it's going to time out. Here is a netstat output while it's trying:

ipfs@ipfs-node:~ $ netstat -tu
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 10.10.2.15:4001         ec2-100-25-248-251:4001 ESTABLISHED
tcp        0    164 10.10.2.15:ssh          10.10.2.16:51276        ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-34-245-56-47.e:4001 ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-54-77-221-96.e:4001 ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-34-244-27-104.:4001 ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-34-211-110-6.u:4001 ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-18-220-145-83.:4001 ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-13-59-235-81.u:4001 ESTABLISHED
tcp        0      0 10.10.2.15:ssh          10.10.2.16:46422        ESTABLISHED
tcp        0      0 localhost:5001          localhost:34362         ESTABLISHED
tcp        0      0 localhost:34362         localhost:5001          ESTABLISHED
tcp        0      0 10.10.2.15:4001         ec2-18-191-132-124:4001 ESTABLISHED

**Comment by fphammerle** [comment]
> Looks like it's going to time out. 

So the webui is not available locally.

I guess either your ipfs config is broken (i.e., no bootstrap nodes configured) or something is preventing your ipfs daemon from connecting to other peers (Check by running `ipfs swarm peers`).

Please also verify that the result of `ipfs bootstrap list` is *not* empty.

**Comment by ThomasFreedman** [comment]
I got results from both of those commands. If config is broken it's b/c that's how it gets installed. One thing that is different from when I installed ipfs before and the webui worked, is that before I used a 3rd party apt repo by siderus. Now, I'm using ipfs installed with ipfs-update, a golang based installation. Here's what those 2 commands produced:

ipfs bootstrap list
/dnsaddr/bootstrap.libp2p.io/ipfs/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN
/dnsaddr/bootstrap.libp2p.io/ipfs/QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa
/dnsaddr/bootstrap.libp2p.io/ipfs/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb
/dnsaddr/bootstrap.libp2p.io/ipfs/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt
/ip4/104.131.131.82/tcp/4001/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ
/ip4/104.236.179.241/tcp/4001/ipfs/QmSoLPppuBtQSGwKDZT2M73ULpjvfd3aZ6ha4oFGL1KrGM
/ip4/104.236.76.40/tcp/4001/ipfs/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64
/ip4/128.199.219.111/tcp/4001/ipfs/QmSoLSafTMBsPKadTEgaXctDQVcqN88CNLHXMkTNwMKPnu
/ip4/178.62.158.247/tcp/4001/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd
/ip6/2400:6180:0:d0::151:6001/tcp/4001/ipfs/QmSoLSafTMBsPKadTEgaXctDQVcqN88CNLHXMkTNwMKPnu
/ip6/2604:a880:1:20::203:d001/tcp/4001/ipfs/QmSoLPppuBtQSGwKDZT2M73ULpjvfd3aZ6ha4oFGL1KrGM
/ip6/2604:a880:800:10::4a:5001/tcp/4001/ipfs/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64
/ip6/2a03:b0c0:0:1010::23:1001/tcp/4001/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd

ipfs swarm peers
/ip4/100.25.248.251/tcp/4001/ipfs/QmWgZLonFZVHbeqLqQUpSkcSaMB1BZ3PNpxAr2ew2paAAE
/ip4/13.59.235.81/tcp/4001/ipfs/QmYfywze6ADxF45Rmmcbv28MZQQ7QuiD1bV88n3Twyq7Ld
/ip4/18.191.132.124/tcp/4001/ipfs/Qmc5NXfUgCfBC8AMUFfDQCP4kjb5ptcAZjq2UT8tEwAGmi
/ip4/18.220.145.83/tcp/4001/ipfs/QmTQcU2VaW3RhCD6brepGvSwZX5X2ArwZjNwZ77yL2QNMS
/ip4/34.211.110.6/tcp/4001/ipfs/QmNmni8sGVa6sqoaUT42tPr9EYuVXTV3rhNW8mnFuzqdA6
/ip4/34.244.27.104/tcp/4001/ipfs/QmQaTKPoYpxecV4rnHrf4aKdgyRrCkuCKRE83DE6msdZg8
/ip4/34.245.56.47/tcp/4001/ipfs/QmPXPHqF2fxHEu8uQUknhRBysA1gmAQDUNqiBaeCpXsvSu
/ip4/54.77.221.96/tcp/4001/ipfs/QmSEwmV67SWoiKsKUiBfCBpTJjxr7ypQS7Md9Nye4EZX8V

**Comment by fphammerle** [comment]
Thanks for running those commands

So your ipfs daemon successfully connects to some peers.
However, the daemon is not able to fetch the webgui
(`ipfs ls QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ` fails)

Unfortunately, I have no idea what might cause this issue.

**Comment by ThomasFreedman** [comment]
Thanks for the step thru trying to resolve the issue with me. This is the "official" origin for the ipfs webui isn't it? If so this thread at least documents a problem. If others should see the same thing it may provide clues as to the cause.

Clearly there is a listener on port 5001, but sounds like the html / css / etc code for the webui is being sourced from a non-local ipfs node which my node can't access for some odd reason.

Makes me wonder a bit about ipfs, since as you say I do have connections with other nodes.

I know ipfs is rapidly evolving, so hopefully this issue will be addressed in due time. 

Thank you again for your help, and all you do to support a free and open internet.

**Comment by ThomasFreedman** [comment]
I don't know what was going on and why I couldn't access the /webui page as described above, it's one of those odd mysteries.

I no longer have that issue, and I have completed my IPFS installation scripts. The most versatile version is a unified installer for 2 classes of platforms:

1. Systems that use the Raspbian OS such the Raspberry Pi3 and Pi3+
2. Any system capable of running Debian OS installed with the Debian net installer

You can find this script and documentation on my github repository: https://github.com/ThomasFreedman/ipfs

All comments welcome.

**Comment by lidel** [comment]
@ThomasFreedman glad it got resolved! Were you still using 0.4.19, or did you update to 0.4.20+?
